File: apps/ios/Moodish/Moodish/Core/Auth/GoogleAuthSession.swift
import AuthenticationServices
import Foundation
import UIKit
import CryptoKit
import Security

/// Drives the native Google login handoff: opens the agent's
/// `/api/auth/google/start?client=mobile` authorize URL in an
/// `ASWebAuthenticationSession`, and the mobile-aware server redirects
/// back to `moodish://auth-callback?token=...` (see services/agent/src/auth.mjs).
@MainActor
final class GoogleAuthSession: NSObject, ASWebAuthenticationPresentationContextProviding {
    private var session: ASWebAuthenticationSession?

    func signIn(authorizeURL: URL, parameter: String = "token") async throws -> String {
        try await withCheckedThrowingContinuation { continuation in
            let session = ASWebAuthenticationSession(url: authorizeURL, callbackURLScheme: "moodish") { callbackURL, error in
                if let error {
                    continuation.resume(throwing: error)
                    return
                }
                guard
                    let callbackURL,
                    let token = URLComponents(url: callbackURL, resolvingAgainstBaseURL: false)?
                        .queryItems?.first(where: { $0.name == parameter })?.value
                else {
                    continuation.resume(throwing: APIError.server(status: 0, message: "Sign-in did not return the expected callback"))
                    return
                }
                continuation.resume(returning: token)
            }
            session.presentationContextProvider = self
            session.prefersEphemeralWebBrowserSession = true
            self.session = session
            if !session.start() { continuation.resume(throwing: APIError.server(status: 0, message: "Could not open sign-in")) }
        }
    }

    func connectSwiggy(api: APIClient) async throws -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else { throw APIError.server(status: 0, message: "Could not start secure login") }
        func base64url(_ data: Data) -> String { data.base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "") }
        let verifier = base64url(Data(bytes))
        let challenge = base64url(Data(SHA256.hash(data: Data(verifier.utf8))))
        let started = try await api.startSwiggy(challenge: challenge)
        guard let url = URL(string: started.authorizationUrl) else { throw APIError.server(status: 0, message: "Invalid login URL") }
        let code = try await signIn(authorizeURL: url, parameter: "code")
        return try await api.exchangeMobile(code: code, verifier: verifier).token
    }

    nonisolated func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        MainActor.assumeIsolated {
            UIApplication.shared.connectedScenes
                .compactMap { ($0 as? UIWindowScene)?.keyWindow }
                .first ?? ASPresentationAnchor()
        }
    }
}

File: apps/moodish_android/lib/core/google_auth_session.dart
import 'package:flutter_web_auth_2/flutter_web_auth_2.dart';
import 'dart:convert';
import 'dart:math';
import 'package:crypto/crypto.dart';
import 'api_client.dart';

class GoogleAuthException implements Exception {
  final String message;
  GoogleAuthException(this.message);
  @override
  String toString() => message;
}

/// Drives the native Google login handoff: opens the agent's
/// `/api/auth/google/start?client=mobile` authorize URL in a Custom Tab
/// (Chrome Custom Tabs on Android, ASWebAuthenticationSession-equivalent),
/// and the mobile-aware server redirects back to
/// moodish://auth-callback?token=... (see services/agent/src/auth.mjs).
class GoogleAuthSession {
  Future<String> connectSwiggy(ApiClient api) async {
    final random = Random.secure();
    String encode(List<int> bytes) => base64UrlEncode(bytes).replaceAll('=', '');
    final verifier = encode(List.generate(32, (_) => random.nextInt(256)));
    final challenge = encode(sha256.convert(utf8.encode(verifier)).bytes);
    final start = await api.swiggyRequest('/api/swiggy/oauth/start', body: {'mobileChallenge': challenge});
    final code = await signIn(Uri.parse(start['authorizationUrl'] as String), parameter: 'code');
    final result = await api.swiggyRequest('/api/auth/mobile/exchange', body: {'code': code, 'verifier': verifier});
    return result['token'] as String;
  }

  Future<String> signIn(Uri authorizeUrl, {String parameter = 'token'}) async {
    final Uri result;
    try {
      result = Uri.parse(
        await FlutterWebAuth2.authenticate(url: authorizeUrl.toString(), callbackUrlScheme: 'moodish'),
      );
    } catch (error) {
      throw GoogleAuthException("Couldn't complete sign-in.");
    }
    final token = result.queryParameters[parameter];
    if (token == null || token.isEmpty) {
      throw GoogleAuthException("Sign-in didn't return the expected callback value");
    }
    return token;
  }
}
